src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,239 @@
+struct LinearIsotropicConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  double conductivity_ = -1.0;
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE T2 operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    SLIC_ASSERT_MSG(conductivity_ > 0.0,
+                    "Conductivity must be positive in the linear isotropic conductor material model.");
+
+    return -1.0 * conductivity_ * temperature_gradient;
+  }
+
+  /**
+   * @brief The density (mass per volume) of the material model
+   *
+   * @tparam T1 type of the position variable
+   * @return The density
+   */
+  template <typename T1>
+  SERAC_HOST_DEVICE double density(T1& /* x */) const
+  {
+    SLIC_ASSERT_MSG(density_ > 0.0, "Density must be positive in the linear isotropic conductor material model.");
+
+    return density_;
+  }
+
+  /**
+   * @brief The specific heat capacity (heat capacity per unit mass) of the material model
+   *
+   * @tparam T1 Type of the position variable
+   * @tparam T2 Type of the temperature variable
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE double specificHeatCapacity(T1& /* x */, T2& /* temperature */) const
+  {
+    SLIC_ASSERT_MSG(specific_heat_capacity_ > 0.0,
+                    "Specific heat capacity must be positive in the linear isotropic conductor material model.");
+
+    return specific_heat_capacity_;
+  }
+};
+
+/**
+ * @brief Linear anisotropic thermal material model
+ *
+ * @tparam dim Spatial dimension
+ */
+template <int dim>
+struct LinearConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  tensor<double, dim, dim> conductivity_ = {-1.0};
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE auto operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    for (int i = 0; i < dim; ++i) {
+      SLIC_ASSERT_MSG(conductivity_(i, i) > 0.0,
+                      "Conductivity tensor must be positive definite in linear conductor material model.");
+    }
```

Comment:
I think having positive diagonal entries is necessary, but not sufficient for a symmetric matrix to be positive-definite, e.g.

```cpp
tensor A = {{1.0, -4.0, -4.0}, {-4.0, 1.0, -40}, {-4.0, -4.0, 1.0}};
tensor x = {1.0, 1.0, 1.0};

x * A * x == -21

```

Reply:
Also, maybe this could be checked elsewhere, as this implementation is putting a conditional + assertion in the most deeply-nested loop of the simulation. Since the conductivity is homogeneous for this material, it only needs to be checked once.

Reply:
Yeah, I wasn't sure the best place to put this. I used `ASSERT` so it wouldn't be compiled under release versions. Maybe a `isValid` method that we can call during the setup phase?

Reply:
I agree that the diagonal check isn't sufficient in the general anisotropic case. Checking for positive determinant would be a nice, dimension-agnostic one-liner. Should there also be a symmetry check?

Reply:
positive determinant also isn't quite right either (e.g. det({{-1, 0, 0}, {0, -1, 0}, {0, 0, 1}}) == 1, but the matrix isn't positive definite). I think I have a routine for closed-form eigenfactorization of symmetric 2x2 and 3x3 matrices that may be useful here (is it safe to assume symmetry for the conductivity?)

Reply:
Ah, you're right on the determinant. I agree checking individual eigenvalues is the right approach. 

Googling, I found a paper [(link)](https://www.inderscienceonline.com/doi/abs/10.1504/IJMSI.2014.067111) that offers a proof that the conductivity tensor must be symmetric for linear conduction, even in the case of an anisotrpic, heterogeneous material. So I think it would be safe to use a symmetric eigenvalue approach, possibly after asserting the tensor the user provides is indeed symmetric.

Reply:
I'd think checking for a Hermitian matrix and using [Sylvester's criterion](https://en.wikipedia.org/wiki/Sylvester%27s_criterion) would work here: all of the leading principal minors must be positive, which only requires extracting submatrices and computing determinants. Computing eigenvalues seems like overkill.

Reply:
I moved these checks to the thermal material constructor. Let me know what you think!